File: frontend/src/lib/analytics.ts
/**
 * Client-side analytics assembled from the math mirror — instant feedback for
 * panels while dragging. Server (/api/analytics/*, order validation) stays
 * authoritative at order time. Mirrors backend/app/services/analytics.py.
 */

import {
  bpPerSetEstimate,
  nakedShortUnits,
  breakevens,
  payoffAtExpiry,
  positionEntryCost,
  positionIv,
  premiumBarrierUnderlying,
  probAboveAtExpiry,
  probTouch,
  structuralMaxLoss,
  terminalEv,
  TRADING_HOURS_PER_YEAR,
  type Leg,
} from "./optionsMath";

export type ClientProbabilities = {
  pProfitExpiry: number;
  breakevens: number[];
  pTouchTp: number | null;
  pTouchSl: number | null;
  tpBarrier: number | null;
  slBarrier: number | null;
  evPerContract: number;
  rewardPerContract: number | null;
  riskPerContract: number | null;
  rr: number | null;
  sigmaUsed: number;
};

export function computeProbabilitiesClient(
  legs: Leg[],
  spot: number,
  hoursToExpiry: number,
  tpPremium: number | null,
  slPremium: number | null,
): ClientProbabilities {
  const tau = Math.max(hoursToExpiry, 0) / TRADING_HOURS_PER_YEAR;
  const sigma = positionIv(legs);
  const entry = positionEntryCost(legs);
  const lo = spot * 0.75;
  const hi = spot * 1.25;
  const bes = breakevens(legs, lo, hi);

  const edges = [lo, ...bes, hi];
  let pProfit = 0;
  for (let i = 0; i < edges.length - 1; i++) {
    const mid = 0.5 * (edges[i] + edges[i + 1]);
    if (payoffAtExpiry(legs, mid) > 0) {
      pProfit +=
        probAboveAtExpiry(spot, edges[i], tau, sigma) -
        probAboveAtExpiry(spot, edges[i + 1], tau, sigma);
    }
  }

  const tauEval = tau / 2;
  let pTp: number | null = null;
  let pSl: number | null = null;
  let tpBarrier: number | null = null;
  let slBarrier: number | null = null;
  if (tpPremium !== null) {
    tpBarrier = premiumBarrierUnderlying(legs, tpPremium, tauEval, lo, hi);
    if (tpBarrier !== null) pTp = probTouch(spot, tpBarrier, tau, sigma);
  }
  if (slPremium !== null) {
    slBarrier = premiumBarrierUnderlying(legs, slPremium, tauEval, lo, hi);
    if (slBarrier !== null) pSl = probTouch(spot, slBarrier, tau, sigma);
  }

  const ev = terminalEv(legs, spot, tau, sigma, tpPremium, slPremium) * 100;
  const reward = tpPremium !== null ? (tpPremium - entry) * 100 : null;
  const risk = slPremium !== null ? (entry - slPremium) * 100 : null;

  return {
    pProfitExpiry: pProfit,
    breakevens: bes.map((b) => Math.round(b * 100) / 100),
    pTouchTp: pTp,
    pTouchSl: pSl,
    tpBarrier,
    slBarrier,
    evPerContract: ev,
    rewardPerContract: reward,
    riskPerContract: risk,
    rr: reward !== null && risk !== null && risk > 0 ? reward / risk : null,
    sigmaUsed: sigma,
  };
}

export type ClientSizing = {
  contracts: number;
  entryCost: number;
  maxLossAtStop: number;
  maxLossStructural: number;
  buyingPowerPct: number;
  perContractRisk: number;
  /** What the stop-risk budget ALONE would allow — when BP binds below this,
   * the gap is the price of undefined risk (naked short margin). */
  riskBudgetContracts: number;
  reasons: string[];
};

export function computeSizingClient(
  legs: Leg[],
  accountEquity: number,
  maxLossPct: number,
  slPremium: number,
  bpCapPct: number,
  spot = 0,
): ClientSizing {
  const reasons: string[] = [];
  const entry = positionEntryCost(legs); // +debit / -credit, per share
  const empty = { contracts: 0, entryCost: 0, maxLossAtStop: 0, maxLossStructural: 0,
                  buyingPowerPct: 0, perContractRisk: 0, riskBudgetContracts: 0 };
  if (Math.abs(entry) < 0.01) {
    return { ...empty, reasons: ["net premium is zero - nothing to size"] };
  }
  // Stop-based risk is sign-agnostic: SL sits below entry on the
  // position-value axis for debit AND credit structures.
  const perSetRisk = Math.max((entry - slPremium) * 100, 0);
  if (perSetRisk <= 0) {
    return { ...empty, reasons: ["stop-loss premium must be below entry"] };
  }
  // Capital per set is the BROKER's number, estimated: debit for longs,
  // structural width for defined risk, cash-secured strike value for
  // uncovered short puts (verified against Alpaca's own cost basis).
  const structural = structuralMaxLoss(legs);
  const perSetCost = spot > 0 ? bpPerSetEstimate(legs, spot)
    : entry > 0 ? entry * 100
    : structural !== null ? structural * 100
    : perSetRisk * 3;
  const perSetStructural =
    structural !== null ? structural * 100 : Math.max(perSetCost, perSetRisk * 3);
  if (structural === null) {
    reasons.push("undefined risk (net short calls) - stop + margin sizing only");
  }
  const budget = accountEquity * maxLossPct;
  const riskBudgetContracts = Math.max(Math.floor(budget / perSetRisk), 0);
  let contracts = riskBudgetContracts;
  if (contracts < 1) {
    reasons.push(`risk/contract $${perSetRisk.toFixed(0)} exceeds budget $${budget.toFixed(0)}`);
  }
  const bpBudget = accountEquity * bpCapPct;
  if (contracts * perSetCost > bpBudget && perSetCost > 0) {
    contracts = Math.min(contracts, Math.floor(bpBudget / perSetCost));
    const nakedPuts = nakedShortUnits(legs, "P");
    if (nakedPuts > 0) {
      reasons.push(
        `BP-bound: uncovered short put is cash-secured (~$${(perSetCost / 1000).toFixed(1)}k/set at the broker). ` +
        `Stop-risk budget alone allows ${riskBudgetContracts} — add a protective put wing to size by stop risk, ` +
        `or raise BP CAP (ACCOUNT tab)`,
      );
    } else {
      reasons.push(`capped by buying-power limit ${(bpCapPct * 100).toFixed(0)}%`);
    }
  }
  return {
    contracts,
    entryCost: contracts * perSetCost,
    maxLossAtStop: contracts * perSetRisk,
    maxLossStructural: contracts * perSetStructural,
    buyingPowerPct: accountEquity ? (contracts * perSetCost) / accountEquity : 0,
    perContractRisk: perSetRisk,
    riskBudgetContracts,
    reasons,
  };
}

File: backend/app/services/analytics.py
"""Position analytics assembled from the math core: sizing, probabilities,
heatmap surfaces. Single source of truth at order time — the TS mirror gives
instant drag feedback, this module gives the authoritative numbers.
"""

from __future__ import annotations

import math
from dataclasses import asdict, dataclass

from app.services.options_math import (
    RISK_FREE,
    TRADING_HOURS_PER_YEAR,
    Leg,
    breakevens,
    payoff_at_expiry,
    position_entry_cost,
    position_pl,
    position_value,
    premium_barrier_underlying,
    prob_above_at_expiry,
    prob_touch,
    structural_max_loss,
    terminal_ev,
)


@dataclass
class SizingResult:
    contracts: int
    entry_cost: float          # dollars, total (per contract-set * contracts)
    max_loss_at_stop: float    # dollars at SL
    max_loss_structural: float # dollars if held to worst case (long: full debit)
    buying_power_pct: float
    per_contract_risk: float
    reasons: list[str]


def position_iv(legs: list[Leg]) -> float:
    """Risk IV for the underlying process: entry-cost-weighted leg IV."""
    weights = [abs(leg.qty * leg.entry) for leg in legs]
    total = sum(weights)
    if total <= 0:
        ivs = [leg.iv for leg in legs if leg.iv > 0]
        return sum(ivs) / len(ivs) if ivs else 0.0
    return sum(w * leg.iv for w, leg in zip(weights, legs)) / total


def compute_sizing(
    legs: list[Leg],
    account_equity: float,
    max_loss_pct: float,
    sl_premium: float,
    bp_cap_pct: float,
) -> SizingResult:
    reasons: list[str] = []
    entry = position_entry_cost(legs)  # per share, one contract-set (+debit / -credit)
    if abs(entry) < 0.01:
        return SizingResult(0, 0, 0, 0, 0, 0, ["net premium is zero - nothing to size"])

    # Risk per set is stop-based and sign-agnostic: SL premium sits below entry
    # on the position-value axis for debit AND credit structures alike.
    per_set_risk = max((entry - sl_premium) * 100, 0.0)
    if per_set_risk <= 0:
        return SizingResult(0, 0, 0, 0, 0, 0, ["stop-loss premium must be below entry"])

    # Capital consumed per set: debit paid for longs; margin (structural max
    # loss) for defined-risk credit; stop-risk proxy for undefined-risk shorts.
    structural = structural_max_loss(legs)
    if entry > 0:
        per_set_cost = entry * 100
        per_set_structural = structural * 100 if structural is not None else entry * 100
    else:
        if structural is None:
            per_set_cost = per_set_risk * 3
            per_set_structural = per_set_cost
            reasons.append("undefined risk (net short calls) - stop-based sizing only")
        else:
            per_set_cost = structural * 100
            per_set_structural = structural * 100

    budget = account_equity * max_loss_pct
    contracts = int(budget // per_set_risk)
    if contracts < 1:
        reasons.append(
            f"risk per contract ${per_set_risk:.0f} exceeds budget ${budget:.0f}"
        )
        contracts = 0

    bp_budget = account_equity * bp_cap_pct
    if contracts * per_set_cost > bp_budget and per_set_cost > 0:
        capped = int(bp_budget // per_set_cost)
        if capped < contracts:
            contracts = capped
            reasons.append(f"capped by buying-power limit {bp_cap_pct:.0%}")

    return SizingResult(
        contracts=contracts,
        entry_cost=round(contracts * per_set_cost, 2),
        max_loss_at_stop=round(contracts * per_set_risk, 2),
        max_loss_structural=round(contracts * per_set_structural, 2),
        buying_power_pct=round(contracts * per_set_cost / account_equity, 4) if account_equity else 0.0,
        per_contract_risk=round(per_set_risk, 2),
        reasons=reasons,
    )


def compute_probabilities(
    legs: list[Leg],
    spot: float,
    hours_to_expiry: float,
    tp_premium: float | None,
    sl_premium: float | None,
) -> dict:
    tau = max(hours_to_expiry, 0.0) / TRADING_HOURS_PER_YEAR
    sigma = position_iv(legs)
    entry = position_entry_cost(legs)

    lo, hi = spot * 0.75, spot * 1.25
    bes = breakevens(legs, lo, hi)

    # P(profit at expiry): sum probability mass over profitable regions.
    edges = [lo] + bes + [hi]
    p_profit = 0.0
    for a, b in zip(edges[:-1], edges[1:]):
        mid = 0.5 * (a + b)
        if payoff_at_expiry(legs, mid) > 0:
            pa = prob_above_at_expiry(spot, a, tau, sigma)
            pb = prob_above_at_expiry(spot, b, tau, sigma)
            p_profit += pa - pb

    # Map TP/SL premiums to underlying barriers at mid-horizon, then use
    # first-passage. Assumption (shown in UI): barrier evaluated at tau/2.
    tau_eval = tau / 2
    p_tp = p_sl = None
    tp_barrier = sl_barrier = None
    if tp_premium is not None:
        tp_barrier = premium_barrier_underlying(legs, tp_premium, tau_eval, lo, hi)
        if tp_barrier is not None:
            p_tp = prob_touch(spot, tp_barrier, tau, sigma)
    if sl_premium is not None:
        sl_barrier = premium_barrier_underlying(legs, sl_premium, tau_eval, lo, hi)
        if sl_barrier is not None:
            p_sl = prob_touch(spot, sl_barrier, tau, sigma)

    ev = terminal_ev(legs, spot, tau, sigma, tp_premium, sl_premium) * 100

    reward = (tp_premium - entry) * 100 if tp_premium is not None else None
    risk = (entry - sl_premium) * 100 if sl_premium is not None else None
    rr = round(reward / risk, 2) if reward and risk and risk > 0 else None

    return {
        "p_profit_expiry": round(p_profit, 4),
        "breakevens": [round(b, 2) for b in bes],
        "p_touch_tp": round(p_tp, 4) if p_tp is not None else None,
        "p_touch_sl": round(p_sl, 4) if p_sl is not None else None,
        "tp_barrier": round(tp_barrier, 2) if tp_barrier is not None else None,
        "sl_barrier": round(sl_barrier, 2) if sl_barrier is not None else None,
        "ev_per_contract": round(ev, 2),
        "reward_per_contract": round(reward, 2) if reward is not None else None,
        "risk_per_contract": round(risk, 2) if risk is not None else None,
        "rr": rr,
        "sigma_used": round(sigma, 4),
        "assumptions": [
            "sticky-strike IV held constant",
            "risk-neutral GBM drift r=%.2f" % RISK_FREE,
            "touch barriers mapped at half-horizon",
            "EV truncated at TP/SL on terminal distribution",
        ],
    }


def compute_heatmap(
    legs: list[Leg],
    spot: float,
    hours_to_expiry: float,
    price_lo: float,
    price_hi: float,
    price_steps: int = 80,
    time_steps: int = 60,
) -> dict:
    """P/L surface: rows = time (now -> expiry), cols = price (lo -> hi).
    Values are dollars per contract-set."""
    price_steps = max(2, min(price_steps, 200))
    time_steps = max(2, min(time_steps, 150))
    prices = [price_lo + (price_hi - price_lo) * i / (price_steps - 1) for i in range(price_steps)]
    hours = [hours_to_expiry * (1 - i / (time_steps - 1)) for i in range(time_steps)]
    grid: list[list[float]] = []
    for h in hours:
        tau = max(h, 0.0) / TRADING_HOURS_PER_YEAR
        row = [round(position_pl(legs, s, tau) * 100, 2) for s in prices]
        grid.append(row)
    flat = [v for row in grid for v in row]
    return {
        "prices": [round(p, 4) for p in prices],
        "hours": [round(h, 4) for h in hours],
        "pl": grid,
        "min_pl": min(flat),
        "max_pl": max(flat),
    }


def legs_from_payload(payload: list[dict]) -> list[Leg]:
    legs = [Leg.from_dict(d) for d in payload]
    if not legs:
        raise ValueError("at least one leg required")
    for leg in legs:
        if leg.right not in ("C", "P"):
            raise ValueError(f"bad right {leg.right!r}")
        if leg.strike <= 0 or leg.qty <= 0 or leg.side not in (1, -1):
            raise ValueError("bad leg parameters")
        if leg.iv <= 0 or not math.isfinite(leg.iv):
            raise ValueError("leg missing IV")
    return legs
